File: learn_android/code/JAVA/LiveDataTest2/app/src/main/java/com/hkc/livedatatest2/MainActivity.java
package com.hkc.livedatatest2;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {
    ViewModelWithLivedata viewModelWithLivedata;
    TextView textView;
    ImageView imageButtonLike,imageButtonDisike;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textView = findViewById(R.id.textView);
        imageButtonLike = findViewById(R.id.imageButtonLike);
        imageButtonDisike = findViewById(R.id.imageButtonDisike);

        viewModelWithLivedata = new ViewModelProvider(this).get(ViewModelWithLivedata.class);
        viewModelWithLivedata.getLikedNumber().observe(this,new Observer<Integer>(){

            @Override
            public void onChanged(Integer integer) {
                textView.setText(String.valueOf(integer));
            }
        });

        imageButtonLike.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                viewModelWithLivedata.addLikedNumber(1);
            }
        });
        imageButtonDisike.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                viewModelWithLivedata.addLikedNumber(-1);
            }
        });


    }
}

File: learn_android/code/CounterDemo/app/src/main/java/com/hkc/counterdemo/MainActivity.kt
package com.hkc.counterdemo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        var number:Int = 0
//        textView.text = "0"
        val viewModel = ViewModelProvider(this).get(MyViewModel::class.java)
        viewModel.number.observe(this, Observer {
            textView.text = it.toString()
        })

        buttonPlus.setOnClickListener {
            viewModel.modifyNumber(1)
//            number++
//            textView.text = (--number).toString()

        }

        buttonMinus.setOnClickListener {
            viewModel.modifyNumber(1)
//            number--
//            textView.text = (++number).toString()

        }

    }
}